runtime/v2/shim.go

```diff
@@ -256,14 +256,35 @@ func makeConnection(ctx context.Context, address string, onClose func()) (_ io.C
 			}
 		}()
 
-		// TODO: figure out how to invoke onCloseWithShimLog callback when shim connection is closed.
-
-		return conn, nil
```

Comment:
I think we should add `WithBlock()` in `gopts` and `context.WithTimeout()` to ensure that the connection is available. 
By default, the grpc-go client connects to server lazily. In current design, shim is not re-connectable. Once the grpc-go connection's state has been changed from READY to IDLE, we should consider to close connection and then call `onCloseFn`, as what we do in ttrpc. 

So, I think we need goroutine to watch the state change with `WaitForStateChange` here.

REF: https://pkg.go.dev/google.golang.org/grpc#ClientConn.WaitForStateChange

Reply:
Sorry for missing the step 2 you mentioned. Yeah. Need a goroutine as watcher.

Reply:
+1 on the `WithBlock`, we actually discussed what opts to add in Maksym's original PR and this came up, but forgot to add here. I'll add a timeout here also as it looks like `AnonReconnectDialer` sets a 100 second timeout for the ttrpc conn. For the monitoring of state though, do we need to? Perhaps I'm not following 

It looks like we explicitly close the ttrpc conn in `*shim.Delete` and `*shim.Close` which would (for ttrpc) [unblock the receiveLoop and then eventually the users callback gets run shortly after](https://github.com/containerd/ttrpc/blob/main/client.go#L295-L299). For GRPC, if we add `WithBlock` we at least know we didn't proceed until the connection is successful, and then we have our wrapper type to run our callback once we close the conn as well. Is your concern if the connection got severed and it wasn't initiated by us, we won't run the callback?

Reply:
Added WithBlock and the timeout for now

Reply:
> Is your concern if the connection got severed and it wasn't initiated by us, we won't run the callback?

Yes. In linux, if the ttrpc-based shim is being killed unexpectedly, for instance, killed by oom-killer, the close callback will cleanup shim. The running container has been re-parented to `pid 1` and only `pid 1` can receive container's SIG_CHILD and exit code. In this case, the container is out of control so that we have to clean it up.
So you can see that the close callback is to clean dead shim.

But in grpc client, it reconnects to server automatically even if the server is down. That is why we need to watch the state of connection and ensure the callback can be called.

Reply:
Okay following now, makes sense why the callback is invoked if the recv loop finishes for the ttrpc conn also and not just a caller initiated `Close`. WaitForStateChange is marked as experimental, but let me screw around with it. Maksym already had a branch that did a polled approach also

Reply:
Feel free to use changes from that branches.
`GetState` uses mutex behind, so it's going to be a bit expensive to poll,
so `WaitForStateChange` should be slightly better for this, though trickier to track transitions (because it's not wait for state, but rather wait for change of current state)

Reply:
The nice bit is it returns the state it transitioned to, so we can either bail if it's Shutdown/Idle or keep going. I'll try and carve out some time in the next hour to update this

Reply:
@fuweid @mxpv Fixed this to monitor the connection status!